modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogCompactionRunner.java

```diff
@@ -0,0 +1,297 @@
+                LOG.info("Catalog compaction is already in progress, skipping (timestamp={})", newLowWatermark.longValue());
+
+                return CompletableFutures.falseCompletedFuture();
+            }
+
+            fut = startCompaction().whenComplete((res, ex) -> {
+                if (ex != null) {
+                    LOG.warn("Catalog compaction has failed (timestamp={})", ex, newLowWatermark.longValue());
+                } else if (LOG.isDebugEnabled()) {
+                    if (res) {
+                        LOG.debug("Catalog compaction completed successfully (timestamp={})", newLowWatermark.longValue());
+                    } else {
+                        LOG.debug("Catalog compaction skipped (timestamp={})", newLowWatermark.longValue());
+                    }
+                }
+            });
+
+            lastRunFuture = fut;
+
+            return fut;
+        });
+    }
+
+    private CompletableFuture<Boolean> startCompaction() {
+        return determineGlobalMinimumRequiredTime()
+                .thenComposeAsync(ts -> {
+                    Catalog catalog = catalogByTsNullable(ts);
+
+                    if (catalog == null) {
+                        return CompletableFutures.falseCompletedFuture();
+                    }
+
+                    Set<String> requiredNodes = requiredNodes(catalog);
+                    List<String> missingNodes = missingNodes(requiredNodes);
+
+                    if (!missingNodes.isEmpty()) {
+                        if (LOG.isDebugEnabled()) {
+                            LOG.debug("Catalog compaction aborted due to missing cluster members (nodes={})", missingNodes);
+                        }
+
+                        return CompletableFutures.falseCompletedFuture();
+                    }
+
+                    return catalogManager.compactCatalog(catalog.time());
+                }, executor);
+    }
+
+    private @Nullable Catalog catalogByTsNullable(long ts) {
+        try {
+            int catalogVer = catalogManager.activeCatalogVersion(ts);
+            return catalogManager.catalog(catalogVer - 1);
+        } catch (IllegalStateException e) {
+            return null;
+        }
+    }
+
+    private CompletableFuture<Long> determineGlobalMinimumRequiredTime() {
+        CatalogMinimumRequiredTimeRequest request = CatalogMinimumRequiredTimeRequestImpl.builder().build();
+        LogicalTopologySnapshot logicalTopology = logicalTopologyService.localLogicalTopology();
+        List<CompletableFuture<?>> ackFutures = new ArrayList<>(logicalTopology.nodes().size());
+        AtomicLong minimumRequiredTimeHolder = new AtomicLong(Long.MAX_VALUE);
+
+        for (LogicalNode node : logicalTopology.nodes()) {
+            CompletableFuture<NetworkMessage> fut = messagingService.invoke(node, request, ANSWER_TIMEOUT)
+                    .whenComplete((msg, e) -> {
+                        long time = ((CatalogMinimumRequiredTimeResponse) msg).timestamp();
```

Comment:
I think this code raise NPE, when messagingService.invoke(node, request, ANSWER_TIMEOUT) fails.

Reply:
Fixed, thanks.